src/enclave/httpsig.h

```diff
@@ -0,0 +1,314 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the Apache 2.0 License.
+#pragma once
+
+#include "httpparser.h"
+#include "node/clientsignatures.h"
+
+#include <fmt/format_header_only.h>
+#include <mbedtls/base64.h>
+#include <optional>
+#include <string>
+#include <tls/keypair.h> // TODO: Only used for hashing
+
+namespace enclave
+{
+  // Implements verification of "Signature" scheme from
+  // https://tools.ietf.org/html/draft-cavage-http-signatures-12
+  //
+  // Tested with RequestClient in tests/infra/clients.py
+  //
+  // TODO:
+  //    - Only supports public key crytography (i.e. no HMAC)
+  //    - Only supports SHA-256 as digest algorithm
+  //    - Only supports ecdsa-sha256 as signature algorithm
+  //    - keyId is ignored
+  class HttpSignatureVerifier
+  {
+  private:
+    static constexpr auto HTTP_HEADER_AUTHORIZATION = "Authorization";
+    static constexpr auto HTTP_HEADER_DIGEST = "Digest";
+
+    static constexpr auto DIGEST_SHA256 = "SHA-256";
+
+    static constexpr auto AUTH_SCHEME = "Signature";
+    static constexpr auto SIGN_PARAMS_KEYID = "keyId";
+    static constexpr auto SIGN_PARAMS_SIGNATURE = "signature";
+    static constexpr auto SIGN_PARAMS_ALGORITHM = "algorithm";
+    static constexpr auto SIGN_PARAMS_HEADERS = "headers";
+    static constexpr auto SIGN_ALGORITHM = "ecdsa-sha256";
+
+    static constexpr auto SIGN_PARAMS_DELIMITER = ",";
+    static constexpr auto SIGN_PARAMS_HEADERS_DELIMITER = " ";
+
+    const http::HeaderMap& headers;
+    const std::vector<uint8_t>& body;
+
+    struct SignatureParams
+    {
+      std::string_view signature = {};
+      std::string_view signature_algorithm = {};
+      std::vector<std::string_view> signed_headers;
+    };
+
+    bool parse_auth_scheme(std::string_view& auth_header_value)
+    {
+      auto next_space = auth_header_value.find(" ");
+      if (next_space == std::string::npos)
+      {
+        LOG_FAIL_FMT("Authorization header only contains one field!");
+        return false;
+      }
+      auto auth_scheme = auth_header_value.substr(0, next_space);
+      if (auth_scheme != AUTH_SCHEME)
+      {
+        LOG_FAIL_FMT("{} is the only supported scheme", AUTH_SCHEME);
+        return false;
+      }
+      auth_header_value = auth_header_value.substr(next_space + 1);
+      return true;
+    }
+
+    // TODO: This should move to a specific base64 encoding-decoding file
+    std::vector<uint8_t> raw_from_b64(const std::string_view& b64_string)
+    {
+      // TODO: Calculate the size of decoded bytes based on size of input base64
+      // string
```

Comment:
`mbedtls_base64_decode` docs say it will give us the required size if we pass `dst = nullptr`.